Spark: Using a Secret for setting AWS credentials in Spark History Server

alembiewski opened the pull request:
This PR introduces the new operator parameter  `awsCredentialsSecretName`, which is the name of a `Secret` for getting AWS credentials for Spark History Server.

Currently, AWS credentials can only be passed to Spark History Server via `spark.hadoop.fs.s3a.*` properties in `historyServerOpts` during the installation. Although this solution will work, it is not optimal because sensitive data must be passed un-encrypted via the command-line.
The solution to this is to store sensitive data in Kubernetes Secrets and then mount values as environment variables to pods.

Changes:
- Introduced `awsCredentialsSecretName` for SHS
- Updated Spark History docs
- Added section describing how to create a Secret with AWS credentials and then use it in Spark Application

Tests can be found here: https://github.com/mesosphere/kudo-spark-operator/pull/88

Commit:
Spark: Using a Secret for setting AWS credentials in Spark History Server

Commit:
Add docs describing integration with S3

akirillov commented on repository/spark/operator/params.yaml:
```diff
@@ -114,6 +114,11 @@ parameters:
     default: ""
     displayName: "Spark History Server Persistent Volume Claim Name"
 
+  - name: awsCredentialsSecretName
```
Is there another way of providing AWS credentials via secret without exposing it in the Operator configuration?

akirillov reviewed (commented):
Thanks, @alembiewski. Overall LGTM but adding an AWS-specific property handling doesn't seem very scalable. How about we take a step back and think about options available for providing cloud-specific and other configs in a more generic way (e.g. by mounting secret containing `spark-defaults.conf`)?

Commit:
Add a functionality to mount Spark config file for Spark History Server

alembiewski commented:
Thanks for the review, @akirillov! Switched this functionality to mounting a secret with `spark-defaults.conf` to SHS pod and changed the test, please take another look: https://github.com/mesosphere/kudo-spark-operator/pull/88

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
+          secretKeyRef:
+            name: <Name of a Secret with AWS credentials>
+            key: AWS_SECRET_ACCESS_KEY
+      # in case when Temporary Security Credentials are used
```
👍

akirillov commented on repository/spark/operator/params.yaml:
```diff
@@ -114,6 +114,11 @@ parameters:
     displayName: "Spark History Server Persistent Volume Claim Name"
 
+  - name: historyServerSparkConfSecret
+    description: "Name of a Secret with Spark config file. The secret should be in the same namespace as an Operator."
```
```suggestion
    description: "Name of a Secret with Spark config file. The secret should be in the same namespace as the Operator."
```

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
 
-Make sure [`specs/spark-application.yaml`](../../../../../specs/spark-application.yaml) has these properties specified under `spec.sparkConf`:
+When submitting Spark Applications to the Operator with Spark History server enabled, you need to provide additional 
+configuration depending on storage type you are using for event logging. 
```
I'd suggest avoiding appeals which use personal pronouns with verbs like "need" or "must" in the documentation. How about we use something like that?
```
When Spark History server is enabled, an additional configuration such as security credentials can be required for accessing the storage backend being used for event logging.
```

alembiewski replied:
yes, looks good, updated

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
+This will deploy a Pod and Service for the `Spark History Server` with the `Spark Event Log` directory configured via the `historyServerFsLogDirectory` parameter. Spark Operator also supports Persistent Volume Claim (PVC) based storage. There is a parameter `historyServerPVCName` to pass the name of the PVC. Make sure that provided PVC should have `ReadWriteMany` [access mode](https://kubernetes. ...
 
-This will deploy a Pod and Service for the `Spark History Server` with the `Spark Event Log` directory configured via the `historyServerFsLogDirectory` parameter. This is an S3 backed storage for event logs. Spark Operator also supports Persistent Volume Claim (PVC) based storage. There is a parameter `historyServerPVCName` to pass the name of the PVC. Make sure that provided PVC should have `Rea ...
+If you want to write and read event logs to/from an S3 bucket, you can do the following:
```
```suggestion
To configure S3 for event log storage it is recommended to store AWS security credentials in a `Secret` in the following way:
```

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
-Make sure [`specs/spark-application.yaml`](../../../../../specs/spark-application.yaml) has these properties specified under `spec.sparkConf`:
+When submitting Spark Applications to the Operator with Spark History server enabled, you need to provide additional 
+configuration depending on storage type you are using for event logging. 
+For S3, an application configuration spec could be the following:
```
```suggestion
The following steps are required for configuring Spark Application event log storage in S3:
```

akirillov commented on repository/spark/operator/params.yaml:
```diff
@@ -114,6 +114,11 @@ parameters:
+  - name: historyServerSparkConfSecret
+    description: "Name of a Secret with Spark config file. The secret should be in the same namespace as an Operator."
+    default: ""
+    displayName: "Name of a Secret with AWS credentials"
```
This looks a little bit misleading because (at least from my understanding) this secret contains the full config for History server,  isn't it? There's another way to configure it using `historyServerOpts` so it looks like `historyServerSparkConfSecret` should be used only for sensitive data.

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
+configuration depending on storage type you are using for event logging. 
+For S3, an application configuration spec could be the following:
 
+1) Create a `Secret` with AWS credentials as described [here](configuration.md#integration-with-aws-s3).
```
```suggestion
1) Create a `Secret` with AWS credentials as described in [Integration with AWS S3](configuration.md#integration-with-aws-s3) Operator documentation.
```

akirillov commented on repository/spark/docs/latest/history-server.md:
```diff
@@ -5,30 +5,75 @@ Spark History Server Configuration
+For S3, an application configuration spec could be the following:
 
+1) Create a `Secret` with AWS credentials as described [here](configuration.md#integration-with-aws-s3).
+2) SparkApplication configuration could be as the following:
```
```suggestion
2) Create a `SparkApplication` configuration with AWS credentials environment variables populated from a `Secret` created at the previous step. Here's an example:
```

akirillov commented on repository/spark/docs/latest/configuration.md:
```diff
@@ -75,3 +75,66 @@ purpose KUDO Spark creates a `Role` for them and binds it to a service account p
+or accessing S3 buckets from Spark workloads. 
+With this method all sensitive data is stored as Kubernetes Secret and mounted to pods via environment variables.
+
+First, you need to convert AWS credentials to base64 as follows:
```
```suggestion
In order to store AWS credentials in a `Secret`, the credentials should be converted to base64 first:
```

akirillov commented on repository/spark/docs/latest/configuration.md:
```diff
@@ -75,3 +75,66 @@ purpose KUDO Spark creates a `Role` for them and binds it to a service account p
+
+### Integration with AWS S3
+This section describes the steps required to provide an access to S3 when using Spark History server 
+or accessing S3 buckets from Spark workloads. 
```
Suggestion:

This section describes the steps required for configuring secure access to S3 buckets for Spark workloads and Spark History server.
or accessing S3 buckets from Spark workloads.

alembiewski replied:
done

akirillov commented on repository/spark/docs/latest/configuration.md:
```diff
@@ -75,3 +75,66 @@ purpose KUDO Spark creates a `Role` for them and binds it to a service account p
+
+To read more about Secrets, refer to the [official K8s documentation](https://kubernetes.io/docs/concepts/configuration/secret/).
+
+Here is the example using a Secret to pass AWS credentials as environment variables to `SparkApplication`: 
```
```suggestion
Here is an example configuration which uses a `Secret` to pass AWS credentials as environment variables to `SparkApplication`: 
```

akirillov commented on repository/spark/docs/latest/configuration.md:
```diff
@@ -75,3 +75,66 @@ purpose KUDO Spark creates a `Role` for them and binds it to a service account p
+          secretKeyRef:
+            name: aws-credentials
+            key: AWS_SECRET_ACCESS_KEY
+      # in case when Temporary Security Credentials are used
```
👍 really like the attention to such minor details

akirillov reviewed (changes requested):
Thanks, @alembiewski!
Really liked the idea and the implementation of the secure access to S3 buckets. It wouldn't harm to massage the docs a bit though.

alembiewski committed:
Apply suggestions from code review

Co-Authored-By: Anton Kirillov <akirillov.mailbox@gmail.com>

alembiewski committed:
Update documentation based on code-review suggestions

Signed-off-by: Alex Lembiewski <alembiyeuski.c@d2iq.com>

akirillov commented on repository/spark/operator/params.yaml:
```diff
@@ -114,6 +114,11 @@ parameters:
+  - name: historyServerSparkConfSecret
+    description: "Name of a Secret with Spark config file. The secret should be in the same namespace as the Operator."
+    default: ""
+    displayName: "Name of a Secret with AWS credentials"
```
```suggestion
    displayName: "Name of a Secret with Spark History Server configuration file"
```

alembiewski replied:
good catch, thanks! applying the suggestion

akirillov reviewed (approved):
Thanks, @alembiewski! LGTM but can we please update `displayName` for `historyServerSparkConfSecret` parameter before merging?

alembiewski committed:
Update "displayName" for "historyServerSparkConfSecret" parameter

Co-Authored-By: Anton Kirillov <akirillov.mailbox@gmail.com>
